# Laboratorio 2

Bienvenido al segundo laboratorio del curso Análisis de Datos en Python. 

Escribe el código correspondiente para contestar cada una de las preguntas. Fíjate muy bien en el nombre de las variables, ya que de lo contrario, tu trabajo no podrá ser evaluado de manera correcta. 

Debes ir ejecutando las celdas de codigo en el orden que aparecen, para evitar tener errores. No elimines celdas!! 

Este trabajo tiene un límite de tiempo que puedes visualizar al inicio de esta pantalla. 

### Pregunta 1
¿Cuál es el promedio y desviación estándar de la cantidad de basura diaria generada en cada calle? 

Indicaciones: 
- Guarda los datos en un dataframe, bajo la variable "basuraxcalle"
- El dataframe debe tener 3 columnas: "Calle", "Cantidad promedio [kilos de basura]" y "Desv est [kilos de basura]" 

In [107]:
import pandas as pd
import numpy as np

ciudad = pd.read_excel('Datos.xlsx',sheet_name = 'Ciudad')
basura = pd.read_excel('Datos.xlsx',sheet_name = 'Histórico de basura')

basuraxcalle= None

# your code here



In [4]:
ciudad.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3240 entries, 0 to 3239
Data columns (total 5 columns):
 #   Column                                                                                                                          Non-Null Count  Dtype  
---  ------                                                                                                                          --------------  -----  
 0   Se considera que la ciudad comienza en la esquina inferior izquierda de la cuadrícula que la representa, y entonces: X=0 e Y=0  3240 non-null   object 
 1   Coordenada (X,Y) donde comienza la calle                                                                                        3240 non-null   object 
 2   Coordenada (X,Y) donde termina la calle                                                                                         3240 non-null   object 
 3   Unnamed: 3                                                                                              

Revisemos que el dataframe haya quedado bien creado, la siguiente celda no tiene puntaje, pero te servirá para que puedas revisar que hayas creado bien tu dataframe 

In [15]:
basura['Calle'].value_counts()

Calle 1       28
Calle 2164    28
Calle 2154    28
Calle 2155    28
Calle 2156    28
              ..
Calle 1084    28
Calle 1085    28
Calle 1086    28
Calle 1087    28
Calle 3240    28
Name: Calle, Length: 3240, dtype: int64

In [167]:
basuraCalculos = basura.groupby('Calle')['Cantidad [kilos de basura]'].agg(['mean', 'std']).reset_index()
basuraCalculos.columns = ['Calle', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]']
basuraxcalle = basuraCalculos
basuraxcalle

Calle  Cantidad promedio [kilos de basura]  \
0        Calle 1                             2.982143   
1       Calle 10                             3.192857   
2      Calle 100                             2.664286   
3     Calle 1000                             2.742857   
4     Calle 1001                             2.896429   
...          ...                                  ...   
3235   Calle 995                             3.032143   
3236   Calle 996                             2.203571   
3237   Calle 997                             2.964286   
3238   Calle 998                             2.978571   
3239   Calle 999                             2.867857   

      Desv est [kilos de basura]  
0                       1.209240  
1                       1.279447  
2                       1.199449  
3                       1.211191  
4                       1.093578  
...                          ...  
3235                    1.172858  
3236                    0.850918  
3237                    1.126520  
3238                    1.203500  
3239                    0.990717  

[3240 rows x 3 columns]

In [184]:
basura

Día       Calle  Cantidad [kilos de basura]
0        1     Calle 1                         4.5
1        1     Calle 2                         1.3
2        1     Calle 3                         1.6
3        1     Calle 4                         3.6
4        1     Calle 5                         2.8
...    ...         ...                         ...
90715   28  Calle 3236                         2.0
90716   28  Calle 3237                         4.2
90717   28  Calle 3238                         1.1
90718   28  Calle 3239                         1.8
90719   28  Calle 3240                         3.9

[90720 rows x 3 columns]

In [168]:
basuraxcalle

Calle  Cantidad promedio [kilos de basura]  \
0        Calle 1                             2.982143   
1       Calle 10                             3.192857   
2      Calle 100                             2.664286   
3     Calle 1000                             2.742857   
4     Calle 1001                             2.896429   
...          ...                                  ...   
3235   Calle 995                             3.032143   
3236   Calle 996                             2.203571   
3237   Calle 997                             2.964286   
3238   Calle 998                             2.978571   
3239   Calle 999                             2.867857   

      Desv est [kilos de basura]  
0                       1.209240  
1                       1.279447  
2                       1.199449  
3                       1.211191  
4                       1.093578  
...                          ...  
3235                    1.172858  
3236                    0.850918  
3237                    1.126520  
3238                    1.203500  
3239                    0.990717  

[3240 rows x 3 columns]

In [21]:
assert list(basuraxcalle.columns) == ['Calle', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]'], "Los nombres de las columnas son incorrectos"

In [22]:
basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]

Calle  Cantidad promedio [kilos de basura]  Desv est [kilos de basura]
0  Calle 1                             2.982143                     1.20924

El resultado para la Calle 1 deberia ser


|Calle   | Cantidad promedio [kilos de basura]	  | Desv est [kilos de basura]  |
|---|---|---|
|Calle 1  | 2.982143  | 1.20924  |

No es necesario que redondees el numero de promedio o desviacion estandar, ni que hagas ninguna operación con el: guardalo en el dataframe tal cual lo entregue el calculo. 

Revisemos que hayas calculado bien el promedio y desviacion para la calle 1:

In [30]:
assert np.round(basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]["Cantidad promedio [kilos de basura]"].values,2) == [[2.98]], "Fallo en el promedio"
assert np.round(basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]["Desv est [kilos de basura]"].values,2) == [[1.21]], "Fallo en la desviacion estandar"

La siguiente celda hace tests ocultos, en el mismo formato del anterior pero para otras calles seleccionadas al azar! No modifiques, muevas ni elimines la celda

### Pregunta 2

Se define como **calle “intensa”** a aquellas calles cuyo promedio diario de basura generada es **estrictamente mayor a 2,92 kilos por día**.

¿Cuántas de las calles serían consideradas intensas (en función de los datos históricos de 28 días)?

Guarda la cantidad de calles intensas en la variable c_intensas, en formato int. 

NO DEBES REDONDEAR NI APROXIMAR EL PROMEDIO PARA HACER LA CLASIFICACION!!

**Hint: usa el dataframe que construiste para la pregunta anterior!**

In [53]:
basuraxcalle

In [43]:
c_intensas=0
c_intensas=basuraxcalle[basuraxcalle['Cantidad promedio [kilos de basura]']>2.92]
c_intensas=len(c_intensas)
c_intensas

1657

In [44]:
c_intensas=0
c_intensas=basuraxcalle[basuraxcalle['Cantidad promedio [kilos de basura]']>2.92].shape[0]
c_intensas

1657

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda

### Pregunta 3

En el dataframe `ciudad` hay datos de las coordenadas donde comienzan y terminan las calles. 

La comuna comienza en la coordenada 0 y termina en la coordenada 40, tanto para X como para Y. Si se divide territorialmente la comuna de la siguiente manera:

#### a. En X, desde la coordenada 0 a la 22 (inclusive) se define Oeste (O)
#### b. En X, desde la coordenada 23 a la 40 (inclusive) se define Este (E)
#### c. En Y, desde la coordenada 0 a la 27 (inclusive) se define Norte (N)
#### d. En Y, desde la coordenada 28 a la 40 (inclusive) se define Sur (S)

Considerando esta división territorial, las calles de la comuna se clasifican como zonas: N-E, N-O, S-E, S-O. 

Puede suceder que alguna calle comience en una zona y termine en otra. Veamos un ejemplo!

Primero renombremos las columnas

In [108]:
ciudad = ciudad.rename(columns = {'Se considera que la ciudad comienza en la esquina inferior izquierda de la cuadrícula que la representa, y entonces: X=0 e Y=0':'Calle',
          'Coordenada (X,Y) donde comienza la calle':'Coordenada inicial',
       'Coordenada (X,Y) donde termina la calle':'Coordenada final'})

Ahora veamos por ejemplo la calle 28

In [72]:
ciudad[ciudad["Calle"] == "Calle 28"]

Calle Coordenada inicial Coordenada final  Unnamed: 3 Unnamed: 4
27  Calle 28             (0,27)           (0,28)         NaN        NaN

In [86]:
ciudad[ciudad["Calle"] == "Calle 28"]

Calle Coordenada inicial Coordenada final Unnamed: 3 Unnamed: 4  \
27  Calle 28            (0, 27)          (1, 27)       None       None   

   Zona inicial Zona final  
27          N-E        N-E

Esta calle comienza en la coordenada (0,27), pero termina en (0,28). Esto quiere decir, que empieza en la zona N-O, pero termina en la zona S-O. Esta es una clase que no se puede clasificar!!

A continuacion, guarda en la variable `sin_clasificar` la cantidad de calles que no pueden ser clasificadas en una zona, es decir, que comiencen en una y terminen en otra. El formato debe ser int

In [109]:
ciudad

Calle Coordenada inicial Coordenada final  Unnamed: 3 Unnamed: 4
0        Calle 1              (0,0)            (0,1)         NaN        NaN
1        Calle 2              (0,1)            (0,2)         NaN        NaN
2        Calle 3              (0,2)            (0,3)         NaN        NaN
3        Calle 4              (0,3)            (0,4)         NaN        NaN
4        Calle 5              (0,4)            (0,5)         NaN        NaN
...          ...                ...              ...         ...        ...
3235  Calle 3236            (35,40)          (36,40)         NaN        NaN
3236  Calle 3237            (36,40)          (37,40)         NaN        NaN
3237  Calle 3238            (37,40)          (38,40)         NaN        NaN
3238  Calle 3239            (38,40)          (39,40)         NaN        NaN
3239  Calle 3240            (39,40)          (40,40)         NaN        NaN

[3240 rows x 5 columns]

In [111]:
import pandas as pd

# Asumiendo que 'ciudad' es tu DataFrame y que las columnas tienen los nombres 'Coordenada inicial' y 'Coordenada final'
# Primero, vamos a definir una función para clasificar cada calle

def clasificar_calle(coord_ini, coord_fin):
    zona_ini = ('O' if coord_ini[0] <= 22 else 'E') + ('-N' if coord_ini[1] <= 27 else '-S')
    zona_fin = ('O' if coord_fin[0] <= 22 else 'E') + ('-N' if coord_fin[1] <= 27 else '-S')
    return zona_ini == zona_fin

# Luego, aplicamos la función a cada fila del DataFrame y guardamos el resultado en una nueva columna
ciudad['misma_zona'] = ciudad.apply(lambda x: clasificar_calle(eval(x['Coordenada inicial']), eval(x['Coordenada final'])), axis=1)

# Finalmente, contamos cuántas calles no están en la misma zona
sin_clasificar = ciudad[~ciudad['misma_zona']].shape[0]
sin_clasificar

82

In [121]:
def clasificar(coord_inicial, coord_final):
    coord_ini = eval(coord_inicial)
    coord_fin = eval(coord_final)

    if coord_ini[0] <= 22:
        zona_ini = 'O' 
    else:
        zona_ini = 'E'  
    
    if coord_ini[1] <= 27:
        zona_ini += 'N'  
    else:
        zona_ini += 'S'  
    

    if coord_fin[0] <= 22:
        zona_fin = 'O'  
    else:
        zona_fin = 'E'  
    
    if coord_fin[1] <= 27:
        zona_fin += 'N'  
    else:
        zona_fin += 'S'  
    
    return zona_ini == zona_fin

clasificacion = []

for index, row in ciudad.iterrows():
    misma_zona = clasificar(row['Coordenada inicial'], row['Coordenada final'])
    clasificacion.append(misma_zona)
ciudad['misma_zona'] = clasificacion
sin_clasificar = len(ciudad[ciudad['misma_zona'] == False])
sin_clasificar=int(sin_clasificar)
sin_clasificar

82

In [116]:
ciudad[ciudad["Calle"] == "Calle 28"]

Calle Coordenada inicial Coordenada final  Unnamed: 3 Unnamed: 4  \
27  Calle 28             (0,27)           (0,28)         NaN        NaN   

    misma_zona  
27       False

In [157]:
sin_clasificar=None
def clasificar(coord_inicial, coord_final):
    coord_ini = eval(coord_inicial)
    coord_fin = eval(coord_final)

    zona_ini = 'N' if coord_ini[1] <= 27 else 'S'
    zona_ini += 'O' if coord_ini[0] <= 22 else 'E'
    
    zona_fin = 'N' if coord_fin[1] <= 27 else 'S'
    zona_fin += 'O' if coord_fin[0] <= 22 else 'E'
    
    return zona_ini == zona_fin

clasificacion = []

for index, row in ciudad.iterrows():
    misma_zona = clasificar(row['Coordenada inicial'], row['Coordenada final'])
    clasificacion.append(misma_zona)
ciudad['misma_zona'] = clasificacion
sin_clasificar = len(ciudad[ciudad['misma_zona'] == False])
sin_clasificar=int(sin_clasificar)
sin_clasificar

82

### Pregunta 4
Considerando la clasificación anterior,
¿cuántas calles quedan clasificadas en cada zona?

Para contestar, genera un dataframe que tenga 2 columnas: Clasificacion y Calle. Dicho df debe ser almacenado en la variable "callesxzona".

Los valores para clasificacion son: "N-E", "N-O", "No clasificado", "S-E", "S-O"

Ejemplo del dataframe: 

|Clasificacion|Calle|
|---|---|
|N-E| 0|
|N-O| 0|
|No clasificado| 0|
|S-E| 0|
|S-O| 0|



In [164]:
def clasificacion(coord_inicial, coord_final):
    coord_ini = eval(coord_inicial)
    coord_fin = eval(coord_final)

    zona_ini = 'N' if coord_ini[1] <= 27 else 'S'
    zona_ini += '-O' if coord_ini[0] <= 22 else '-E'
    
    zona_fin = 'N' if coord_fin[1] <= 27 else 'S'
    zona_fin += '-O' if coord_fin[0] <= 22 else '-E'
    
    if zona_ini == zona_fin:
        return zona_ini  
    else:
        return 'No clasificado' 

ciudad['Clasificacion'] = ciudad.apply(lambda row: clasificacion(row['Coordenada inicial'], row['Coordenada final']), axis=1)
callesxzona = ciudad.groupby('Clasificacion').size().reset_index(name='Calle')
orden_clasificacion = ["N-E", "N-O", "No clasificado", "S-E", "S-O"]
callesxzona['Clasificacion'] = pd.Categorical(callesxzona['Clasificacion'], categories=orden_clasificacion, ordered=True)
callesxzona = callesxzona.sort_values('Clasificacion').reset_index(drop=True)
callesxzona


Clasificacion  Calle
0             N-E    962
1             N-O   1210
2  No clasificado     82
3             S-E    437
4             S-O    549

In [170]:
ciudad

Calle Coordenada inicial Coordenada final  Unnamed: 3 Unnamed: 4  \
0        Calle 1              (0,0)            (0,1)         NaN        NaN   
1        Calle 2              (0,1)            (0,2)         NaN        NaN   
2        Calle 3              (0,2)            (0,3)         NaN        NaN   
3        Calle 4              (0,3)            (0,4)         NaN        NaN   
4        Calle 5              (0,4)            (0,5)         NaN        NaN   
...          ...                ...              ...         ...        ...   
3235  Calle 3236            (35,40)          (36,40)         NaN        NaN   
3236  Calle 3237            (36,40)          (37,40)         NaN        NaN   
3237  Calle 3238            (37,40)          (38,40)         NaN        NaN   
3238  Calle 3239            (38,40)          (39,40)         NaN        NaN   
3239  Calle 3240            (39,40)          (40,40)         NaN        NaN   

      misma_zona Clasificacion  
0           True           N-O  
1           True           N-O  
2           True           N-O  
3           True           N-O  
4           True           N-O  
...          ...           ...  
3235        True           S-E  
3236        True           S-E  
3237        True           S-E  
3238        True           S-E  
3239        True           S-E  

[3240 rows x 7 columns]

In [ ]:
callesxzona=None
# your code here
raise NotImplementedError

Revisa que hayas creado bien el dataframe!

In [159]:
print(list(callesxzona.columns) == ['Clasificacion', 'Calle'])

clasificaciones = list(callesxzona["Clasificacion"].values)
clasificaciones.sort()

print(clasificaciones == ['N-E', 'N-O', 'No clasificado', 'S-E', 'S-O'])

True
True


Comenzamos con un test visible!

In [160]:
assert callesxzona[callesxzona['Clasificacion'] == 'N-E']["Calle"].values == [[962]]

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda

### Pregunta 5 
¿Cuál es el promedio y desviación estándar de la cantidad de kilos de basura generada por cada zona?

Excluya las calles que quedan sin clasificar. 

Guarde su respuesta en un dataframe con la siguiente estructura: 

|Zona   | Cantidad promedio [kilos de basura]	  | Desv est [kilos de basura]  |
|---|---|---|
|N-O  | 0  | 0  |

Guarde el dataframe en la variable `zona_stats`

In [169]:
basuraxcalle

Calle  Cantidad promedio [kilos de basura]  \
0        Calle 1                             2.982143   
1       Calle 10                             3.192857   
2      Calle 100                             2.664286   
3     Calle 1000                             2.742857   
4     Calle 1001                             2.896429   
...          ...                                  ...   
3235   Calle 995                             3.032143   
3236   Calle 996                             2.203571   
3237   Calle 997                             2.964286   
3238   Calle 998                             2.978571   
3239   Calle 999                             2.867857   

      Desv est [kilos de basura]  
0                       1.209240  
1                       1.279447  
2                       1.199449  
3                       1.211191  
4                       1.093578  
...                          ...  
3235                    1.172858  
3236                    0.850918  
3237                    1.126520  
3238                    1.203500  
3239                    0.990717  

[3240 rows x 3 columns]

In [193]:
ciudadSinNoClasificado = ciudad[ciudad['Clasificacion'] != 'No clasificado']
df = pd.merge(ciudadSinNoClasificado, basuraxcalle, on='Calle')
dfAgrupado = df.groupby('Clasificacion')['Cantidad promedio [kilos de basura]'].agg(['mean', 'std']).reset_index()   
                
dfAgrupado.columns = ['Zona', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]']
zona_stats = dfAgrupado
zona_stats


Zona  Cantidad promedio [kilos de basura]  Desv est [kilos de basura]
0  N-E                             2.926032                    0.207145
1  N-O                             2.920708                    0.206043
2  S-E                             2.931636                    0.222461
3  S-O                             2.928220                    0.209070

In [195]:
ciudadSinNoClasificado = ciudad[ciudad['Clasificacion'] != 'No clasificado']
df = pd.merge(ciudadSinNoClasificado, basura, on='Calle')
dfAgrupado = df.groupby('Clasificacion')['Cantidad [kilos de basura]'].agg(['mean', 'std']).reset_index()

dfAgrupado.columns = ['Zona', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]']
zona_stats = dfAgrupado
zona_stats

Zona  Cantidad promedio [kilos de basura]  Desv est [kilos de basura]
0  N-E                             2.926032                    1.122461
1  N-O                             2.920708                    1.123545
2  S-E                             2.931636                    1.116296
3  S-O                             2.928220                    1.117884

In [176]:
zona_stats

Zona  Cantidad promedio [kilos de basura]  Desv est [kilos de basura]
0  N-E                             2.926032                    0.207145
1  N-O                             2.920708                    0.206043
2  S-E                             2.931636                    0.222461
3  S-O                             2.928220                    0.209070

In [ ]:
zona_stats=None
# your code here
raise NotImplementedError

Revisemos el formato

In [182]:
print(list(zona_stats.columns) == ['Zona', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]'])

cols = list(zona_stats['Zona'].values)
cols.sort()

print(cols == ['N-E', 'N-O', 'S-E', 'S-O'])

True
True


Comenzamos con un test visible:

In [180]:
ans_NE = np.round(zona_stats[zona_stats["Zona"]=='N-E'][["Cantidad promedio [kilos de basura]", "Desv est [kilos de basura]"]].values,3)

assert (ans_NE== [[2.926, 1.122]]).all()

AssertionError: 

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda